jsonschema/src/compilation/mod.rs

```diff
@@ -569,4 +582,58 @@ mod tests {
         assert!(compiled.validate(&instance_ok).is_err());
         assert!(!compiled.is_valid(&instance_ok));
     }
+
+    #[test]
+    fn custom_keyword_with_inner_state() {
+        // Define a custom keyword validator that wraps "minimum"
+        // but maintains a counter of how many times the validator was applied.
+        struct CountingValidator {
+            count: u32,
+        }
+        impl<'instance> CustomKeywordValidator<'instance, '_> for CountingValidator {
+            fn validate(
+                &mut self,
+                instance: &'instance Value,
+                instance_path: JSONPointer,
+                subschema: Arc<Value>,
+                subschema_path: JSONPointer,
+                schema: Arc<Value>,
+            ) -> ErrorIterator<'instance> {
+                let amount = match &*subschema {
+                    Value::Number(x) => x.as_i64().expect("countme value must be integer"),
+                    _ => panic!("Validator requires numeric values"),
+                };
+                self.count += amount as u32;
+                Box::new(None.into_iter())
+            }
+
+            fn is_valid(&mut self, instance: &Value, subschema: &Value, schema: &Value) -> bool {
+                let amount = match subschema {
+                    Value::Number(x) => x.as_i64().expect("countme value must be integer"),
+                    _ => return false,
+                };
+                self.count += amount as u32;
+                true
+            }
+        }
+
+        // define compilation options that include the custom format and the overridden keyword
+        let validator = Arc::new(Mutex::new(Box::new(CountingValidator { count: 0 })));
+        let mut options = JSONSchema::options();
+        let options = options.with_custom_keyword("countme", validator.clone());
```

Comment:
@Stranger6667 unfortunately I think I've hit the end of my expertise with trait objects, Arc, Mutex, Box, etc. Everything else compiles, however I can't figure out how to maintain the type of `validator` here so that I can access the `count` member later on in assertions.

```
error[E0308]: mismatched types
   --> src/compilation/mod.rs:623:62
    |
623 |         let options = options.with_custom_keyword("countme", validator.clone());
    |                               -------------------            ^^^^^^^^^^^^^^^^^ expected `Arc<Mutex<Box<...>>>`, found `Arc<Mutex<Box<CountingValidator>>>`
    |                               |
    |                               arguments to this method are incorrect
    |
    = note: expected struct `Arc<std::sync::Mutex<Box<(dyn for<'instance, 'schema> CustomKeywordValidator<'instance, 'schema> + 'static)>>>`
               found struct `Arc<std::sync::Mutex<Box<CountingValidator>>>`
    = help: `CountingValidator` implements `CustomKeywordValidator` so you could box the found value and coerce it to the trait object `Box<dyn CustomKeywordValidator>`, you will have to change the expected type as well```
    
Also, I'm not thrilled with how the `with_custom_keyword` interface requires an Arc of a Mutex of a Box of the thing... for non inner state use cases that's unnecessary (see the other custom keyword tests above), but I'm not sure how to support both. Any ideas for either problem?

Reply:
Looking at it right now. I'll push a few commits to your branch which will this work and then we can continue discussing the interface :)

Reply:
Here are a few thoughts:

Given the inner state, I think that `custom_keywords` should store initialization functions (boxed this time, unlike `formats`), so such validators won't be shared across multiple locations in the schema. This will remove the need for `Arc<Mutex<T>>`, so no unconditional locking when any custom validator is used - stateless validators will be more performant. Alternatively, we can store something that implements `Validator + Default`, which is likely a better option.

Reply:
Hey, so, I think again my inexperience is showing. My attempts with Validator + Default fail because that cannot be a trait object, and if we try to store a fn in CompilationOptions we can't capture variables when in the closure and that wouldn't be able to clone the `count` in the inner state test, and if we try to store a Fn then that's either not thread safe or not cloneable. Any ideas?

Reply:
Oh, I haven't thought about the trait object thing. Let me play with the code a bit

P.S. Offtopic - I've put together some ideas about this library's `1.0` API and would appreciate any feedback (see https://github.com/Stranger6667/jsonschema-rs/discussions/475)

Reply:
Done in a240bd5aae1903b00328a444046dc7b008793928! However, there are a few issues:

- Initialization functions require `&'static` so, it is problematic to pass a value there (like a counter) and then read from it without leaking.
- The API does not spark joy :( However, given that there could be multiple occurrences of the same keyword, every one of them should have its separate state, so it is probably the only way to provide this kind of capability. I.e. - each validator is unique, so we need to give a way to initialize each of them. With stateless validators, things could be simpler.

Not 100% sure if it is worth the trouble and if the use case I mentioned above could be implemented differently